### PR TITLE
fix: allow only credhub and uaa for apps

### LINKED ISSUE #304: Remove security group that allows apps to communicate with internal network
**Describe the bug**

Replaces https://github.com/SUSE/kubecf/issues/297.

The pod and service CIDRs are currently required to created an app security group that allows connections to the internal k8s network. It allows connection to service brokers like `credhub`.

**To Reproduce**

Always.

**Expected behavior**

This approach is erroneous and we should allow only the `credhub` IP.
An auto-errand can take care of it by resolving `credhub.service.cf.internal` and creating a security group allowing apps to communicate with this IP. We also should add `features.credhub.enabled=true` as default, but remove `credhub` and the auto-errand when this setting is `false`, allowing the usage of an external `credhub` instance.

**Environment**

Kubecf 0.1, all platforms.

### PR DESCRIPTION
**Pull request by f0rmiga**
## Description

Removes the internal security group that allows apps to communicate to the k8s service and pod IP ranges. Instead, create internal security groups allowing apps to communicate with credhub and UAA only.

## Motivation and Context

We currently have some internal app security group stuff for CredHub (from #281); this requires the administrator to set `kube.service_cluster_ip_range` and `kube.pod_cluster_ip_range` in their helm values. This is unwieldy and error-prone; we should instead auto-detect the appropriate values and set the application security groups from code.

This fixes #304.

## How Has This Been Tested?

Manually, running CATS and asserting that the credhub suite pass.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code has security implications.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml [hunk 1]
```diff
@@ -24,16 +25,56 @@
...
+            args:
+            - -c
+            - |-
+              {{- .Files.Get "assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh" | nindent 14 }}
+        envs:
+        - name: CF_CLI_PATH
+          value: /var/vcap/packages/cf-cli-6-linux/bin
+        - name: CF_API
+          value: https://api.((system_domain))
+        - name: CF_USERNAME
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml]: This should be done with a client credential instead (so if the admin ever changes the password restarting the job won't be a problem).

File: deploy/helm/kubecf/assets/operations/instance_groups/uaa.yaml [hunk 2]
```diff
@@ -31,6 +31,49 @@
...
+            args:
+            - -c
+            - |-
+              {{- .Files.Get "assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh" | nindent 14 }}
+        envs:
+        - name: CF_CLI_PATH
+          value: /var/vcap/packages/cf-cli-6-linux/bin
+        - name: CF_API
+          value: https://api.((system_domain))
+        - name: CF_USERNAME
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/operations/instance_groups/uaa.yaml]: Same as above.

File: deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml [hunk 3]
```diff
@@ -24,16 +25,56 @@
...
+            executable: /bin/bash
+            args:
+            - -c
+            - |-
+              {{- .Files.Get "assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh" | nindent 14 }}
+        envs:
+        - name: CF_CLI_PATH
+          value: /var/vcap/packages/cf-cli-6-linux/bin
+        - name: CF_API
+          value: https://api.((system_domain))
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml]: Should we be using the internal host name for CC for this? (Mostly thinking about things like "what if the external endpoint is behind a load balancer we can't reach from the inside", and "which SSL cert is that going to use")

File: deploy/helm/kubecf/assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
+set -o errexit -o nounset -o xtrace
+
+export PATH="${CF_CLI_PATH}:${PATH}"
+
+until curl --insecure --fail --silent --head "${CF_API}"; do
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh]:
```suggestion
until curl --insecure --fail --silent --head "${CF_API}/v2/info"; do
```
We should also add a comment on _why_ poking at the API before UAA is ready is fine.

File: deploy/helm/kubecf/assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    sleep 1
+done
+
+cf api --skip-ssl-validation "${CF_API}"
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh]: Can we do anything to avoid the `--skip-ssl-validation`?

Comment (viovanov) [reply deploy/helm/kubecf/assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh]: +1, we can't leave insecure in 
also `--silent` is a bad idea

File: deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml [hunk 6]
```diff
@@ -24,16 +25,56 @@
...
           registration_interval: 10s
           tags:
             component: credhub
           uris:
           - credhub.((system_domain))
-{{- end }}
 
-{{- if not .Values.features.credhub.enabled }}
+- type: replace
+  path: /instance_groups/name=credhub/jobs/-
```
Comment (viovanov) [inline deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml]: You're adding a container to a pod with an entrypoint that's not long running afaict.

Comment (f0rmiga) [reply deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml]: The idea was to also have a `trap` to clean up the security group on `EXIT`.
As we discussed on a chat, I'll change this to a job that is aware of the replicas.

File: deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml (hunk 3 above)
Comment (f0rmiga) [reply deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml]: Hmm. I'll try to connect to the internal hostname and load the CA cert to be able to remove `--skip-ssl-validation` from the `cf api` command.

File: deploy/helm/kubecf/assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh (hunk 5 above)
Comment (f0rmiga) [reply deploy/helm/kubecf/assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh]: @viovanov I guess the `--silent` is regarding the `curl` command. That command will be triggered multiple times, and we don't want it to spam logs.

For the `--skip-ssl-validation`, I have a plan for it that ties with the commend above about using the internal hostname.

File: deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml (hunk 6 above)
Comment (f0rmiga) [reply deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml]: After re-assessing with Vlad, doing it as s job brings other complexities, like adding a service that performs load balancing to the replicas and then replacing the BOSH DNS aliases. That is a direction we don't want to take at this point. I'll follow my original idea here.

Comment (mook-as) [reply deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml]: :+1: — but I'm still exploring it as a job (`post-start` + `drain` scripts), using the same `POD_IP` mechanism you're using, to see how it goes.  That's _without_ the load balancing part — it'll just be co-located with each uaa / credhub instance.

File: deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml (hunk 3 above)
Comment (f0rmiga) [reply deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml]: Created https://github.com/cloudfoundry-incubator/kubecf/issues/449.

File: deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml (hunk 1 above)
Comment (f0rmiga) [reply deploy/helm/kubecf/assets/operations/instance_groups/credhub.yaml]: Created https://github.com/cloudfoundry-incubator/kubecf/issues/450.

File: deploy/helm/kubecf/assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh (hunk 5 above)
Comment (f0rmiga) [reply deploy/helm/kubecf/assets/scripts/jobs/cf-cli-6-linux/setup_internal_security_group.sh]: Created https://github.com/cloudfoundry-incubator/kubecf/issues/451.

**Pull request merged**